upf/quantifiers_remover.py

```diff
@@ -0,0 +1,174 @@
+        new_problem = self._create_problem_copy_without_actions_and_goals()
+        for a in self._problem.actions().values():
+            na = self._action_without_quantifiers(a)
+            new_problem.add_action(na)
+            self._action_mapping[na] = a
+        for g in self._problem.goals():
+            ng = self._expression_quantifier_remover.remove_quantifiers(g)
+            new_problem.add_goal(ng)
+        return new_problem
+
+    def _create_problem_copy_without_actions_and_goals(self):
+        '''Creates the shallow copy of a problem, without adding the actions
+        with quantifiers and by pushing them to the stack
+        '''
+        new_problem: Problem = Problem("noquantifiers_" + str(self._problem.name()), self._env)
+        for f in self._problem.fluents().values():
+            new_problem.add_fluent(f)
+        for o in self._problem.all_objects():
+            new_problem.add_object(o)
+        for fl, v in self._problem.initial_values().items():
+            new_problem.set_initial_value(fl, v)
+        return new_problem
+
+    def _action_without_quantifiers(self, action) -> Action:
+        #emulates a do-while loop: searching for an available name
+        is_unavailable_name = True
+        while is_unavailable_name:
+            new_action_name = action.name()+ "_" +str(self._counter)
+            self._counter = self._counter + 1
+            is_unavailable_name = self._problem.has_action(new_action_name)
+        new_parameters = OrderedDict()
+        for ap in action.parameters():
+            new_parameters[ap.name()] = ap.type()
+        new_action = Action(new_action_name, new_parameters, self._env)
+
+        for p in action.preconditions():
+            np = self._expression_quantifier_remover.remove_quantifiers(p)
+            new_action.add_precondition(np)
+        for e in action.effects():
+            if e.is_conditional():
+                nc = self._expression_quantifier_remover.remove_quantifiers(e.condition())
+            else:
+                nc = self._env.expression_manager.TRUE()
+            nv = self._expression_quantifier_remover.remove_quantifiers(e.value())
+            ne = Effect(e.fluent(), nv, nc, e.kind())
+            new_action._add_effect_instance(ne)
+        return new_action
+
+    def rewrite_back_plan(self, unconditional_sequential_plan: SequentialPlan) -> SequentialPlan:
+        '''Takes the sequential plan of the problem (created with
+        the method "self.get_rewritten_problem()" and translates the plan back
+        to be a plan of the original problem.'''
+        uncond_actions = unconditional_sequential_plan.actions()
+        cond_actions = []
+        for ai in uncond_actions:
+            if ai.action() in self._action_mapping:
+                cond_actions.append(self._new_action_instance_original_name(ai))
+            else:
+                cond_actions.append(ai)
```

Comment:
Is this case even possible?

Reply:
Yes because the _action_mapping contains only the actions that had a quantifier.